typing/rec_check.ml

```diff
@@ -774,18 +774,77 @@ let rec expression : Typedtree.expression -> term_judg =
         path pth << Dereference;
         list field fields << Dereference;
       ]
-    | Texp_function { cases } ->
+    | Texp_function (params, body) ->
       (*
-         (Gi; _ |- pi -> ei : m[Delay])^i
-         --------------------------------------
-         sum(Gi)^i |- function (pi -> ei)^i : m
+         Function_cases:
+
+         (Gi; _ |- pi -> ei : m[Delay])^i    (**)
+         (Hj    |-{def}  Pj : m[Delay])^j
+         G  := sum(Gi)^i
+         H  := sum(Hj)^j
+         ps := sum(pat(Pj))^j
+         ---------------------------------------------------------
+         G + H - ps |- fun (Pj)^j -> function (pi -> ei)^i : m
+
+         Function_body:
+
+         G |- e : m[Delay]
+         (Hj |-{def}  Pj : m[Delay])^j
+         H  := sum(Hj)^j
+         ps := sum(pat(Pj))^j
+         -------------------------------------
+         G + H - ps |- fun (Pj)^j -> e : m
+
+         (**) Contrarily to match, the values that are pattern-matched
+              are bound locally, so the pattern modes do not influence
+              the final environment.
+      *)
+      let param_pat param =
+        (* param P ::=
+            | ?(pat = expr)
+            | pat
+
+          Define pat(P) as
+              pat if P = ?(pat = expr)
+              pat if P = pat
+          *)
+        match param.fp_kind with
+        | Tparam_pat pat -> pat
+        | Tparam_optional_default (pat, _) -> pat
+      in
+      (* Optional argument defaults.
 
-         Contrarily to match, the value that is pattern-matched
-         is bound locally, so the pattern modes do not influence
-         the final environment.
+         G |-{def} P : m
       *)
-      let case_env c m = fst (case c m) in
-      list case_env cases << Delay
+      let param_default param =
+        match param.fp_kind with
+        | Tparam_optional_default (_, default) ->
+          (*
+             G |- e : m
+             ------------------
+             G |-{def} ?(p=e) : m
+          *)
+            expression default
+        | Tparam_pat _ ->
+          (*
+             ------------------
+             . |-{def} p : m
+          *)
+            empty
+      in
+      let patterns = List.map param_pat params in
+      let defaults = List.map param_default params in
+      let body =
+        match body with
+        | Tfunction_body body -> expression body
+        | Tfunction_cases { cases; _ } ->
+            List.map (fun c mode -> fst (case c mode)) cases
+            |> join
+      in
+      let f = join (body :: defaults) << Delay in
+      (fun m ->
+         let env = f m in
+         remove_patlist patterns env)
```

Comment:
Indeed I hadn't reviewed the rec_check code. I just did and here are my thoughts:

1. The new code is correct. Merging it as-is would be perfectly acceptable. What follows below are merely refactoring suggestions.
2. It is a bit long to my taste. I like it when all cases fit a screen, and try to move longer logic to auxiliary functions.
3. The two inference rules given (thanks for maintaining this!) are a bit redundant. I would factor them out by introducing a specialized judgment for the new `function_body` type.
4. This idea also suggests how to factorize the code: introduce a new `function_body` function at the level of the recursive nest of functions (one per grammatical category), and focus the `Texp_function` code on handling the params and calling `function_body` for the body.
5. I *think* that the params could be handled more nicely as a `fold`. For each param, remove its pattern from the environment and add the default value (if any) under a Delay context.

Reply:
(I'm impressed that @ncik-roberts managed to adapt this code -- whose logic is non-trivial and whose style is rather different from the rest of the codebase -- and @goldfirere managed to review the changes, both correctly.)

Reply:
I took the suggestion to introduce a specialized judgment for `function_body` and factored it out to another function. Thanks for the suggestion, I like this more.

I found it somewhat less readable to convey the `fold` in the inference rules, so I prefer to keep it as-is (and then to keep the code as-is to match the structure of the rule). But, if you feel strongly about this, I can revisit.

Reply:
Thanks! I am fine with the new version. (I think that if I had written the code myself I wouldn't have bothered defining in comments all the auxiliary notations used in the inference rules, and just let the code speak for those, at the benefit of having less code. But both approaches are merit and I am fine with the current state.)